src/main/java/org/truffleruby/core/regexp/RegexpNodes.java

```diff
@@ -405,11 +431,14 @@ public int hash(DynamicObject regexp) {
     public abstract static class MatchStartNode extends CoreMethodArrayArgumentsNode {
 
         @Child private CallDispatchHeadNode dupNode = CallDispatchHeadNode.create();
+        @Child private MatchNode matchNode = MatchNode.create();
 
         @Specialization(guards = "isRubyString(string)")
         public Object matchStart(VirtualFrame frame, DynamicObject regexp, DynamicObject string, int startPos) {
             final DynamicObject dupedString = (DynamicObject) dupNode.call(frame, string, "dup");
-            final Object matchResult = matchCommon(getContext(), this, regexp, dupedString, startPos, true);
+            final Matcher matcher = createMatcher(getContext(), regexp, string, true);
+            int range = StringOperations.rope(string).byteLength();
+            final Object matchResult = matchNode.execute(regexp, dupedString, matcher, startPos, range, true);
 
             if (RubyGuards.isRubyMatchData(matchResult) && Layouts.MATCH_DATA.getRegion((DynamicObject) matchResult).numRegs > 0
                 && Layouts.MATCH_DATA.getRegion((DynamicObject) matchResult).beg[0] == startPos) {
```

Comment:
This last test should be redundant with `onlyMatchAtStart`, right?
Then could it be just asserted?
Is `numRegs > 0` needed? Otherwise we could just return the  matchResult.